### PR TITLE
Fixes GPU memory polling using environment variable filtering

### PR DESCRIPTION
**Pull request by ericleonardis**
### Description
- Fixes bug where nvidia-smi memory poling doesn’t respect visibility masking. 
- Auto-detected GPU now picks from the masked GPUs listed in os.environ[‘CUDA_VISIBLE_DEVICES’]. 
- Tests added, but it is possible that the test might not be included due to cpu only during test. 

### Types of changes

- [x] Bugfix
- [ ] New feature
- [ ] Refactor / Code style update (no logical changes)
- [ ] Build / CI changes
- [ ] Documentation Update
- [ ] Other (explain)

### Does this address any currently open issues?
Addresses Issue #1028  

### Outside contributors checklist

- [ ] Review the [guidelines for contributing](https://github.com/talmolab/sleap/blob/develop/docs/CONTRIBUTING.md) to this repository
- [ ] Read and sign the [CLA](https://github.com/talmolab/sleap/blob/develop/sleap-cla.pdf) and add yourself to the [authors list](https://github.com/talmolab/sleap/blob/develop/AUTHORS)
- [ ] Make sure you are making a pull request against the **develop** branch (not *main*). Also you should start *your branch* off *develop*
- [ ] Add tests that prove your fix is effective or that your feature works
- [ ] Add necessary documentation (if appropriate)

#### Thank you for contributing to SLEAP!
:heart:

### CODE REVIEW

**Review by roomrys — CHANGES_REQUESTED** [review]

File: tests/nn/test_system.py [hunk 1]
```diff
@@ -5,8 +5,37 @@
...
+
+        gpu_memory = get_gpu_memory()
+
+        assert len(gpu_memory) > 0
+        assert len(gpu_memory) == 1
+
+        os.environ.pop("CUDA_VISIBLE_DEVICES", None)
+
+    elif "," in cuda_visible_devices:
+        os.environ["CUDA_VISIBLE_DEVICES"] = "0,1"
```
[the remark this answers is not in the document]
Comment (ericleonardis) [reply tests/nn/test_system.py:36]: it did indeed fail! I have written a new test that actually works on a system with 2 gpus.

File: sleap/nn/system.py [hunk 2]
```diff
@@ -196,12 +198,19 @@ def get_gpu_memory() -> List[int]:
...
+
     command = [
         "nvidia-smi",
         "--query-gpu=memory.free",
         "--format=csv",
     ]
 
+    if "CUDA_VISIBLE_DEVICES" in os.environ.keys():
+        command.append("-i")
+        command.append(os.environ["CUDA_VISIBLE_DEVICES"])
```
[the remark this answers is not in the document]
Comment (ericleonardis) [reply sleap/nn/system.py]: thanks for the suggestion! i wasn't able to get it to work that well, so i tried an alternative method which i just pushed

File: tests/nn/test_system.py [hunk 3]
```diff
@@ -5,8 +5,39 @@
...
+import os
+import pytest
 
 
 def test_get_gpu_memory():
     # Make sure this doesn't throw an exception
     memory = get_gpu_memory()
+
+
+def test_get_gpu_memory_visible():
```
Comment (roomrys) [inline tests/nn/test_system.py:17]: Could we also add a test (possibly it's own function/test) to check that `sleap.nn.system.get_all_gpus` returns the GPUs in the same order (ascending gpu index) as `nvidia_smi` query and same length as `sleap.nn.system.get_gpu_memory`

**Review by roomrys — CHANGES_REQUESTED** [review]
Implementation looks good. Let's just clean-up/solidify the tests. Thanks!

File: tests/nn/test_system.py [hunk 4]
```diff
@@ -5,8 +5,39 @@
...
+
+    cuda_visible_devices = os.environ.get("CUDA_VISIBLE_DEVICES", None)
+
+    if cuda_visible_devices is None or cuda_visible_devices is []:
+        pytest.skip("CUDA_VISIBLE_DEVICES not set.")
+
+    elif cuda_visible_devices == "0":
+        os.environ["CUDA_VISIBLE_DEVICES"] = "0"
```
[the remark this answers is not in the document]
Comment (ericleonardis) [reply tests/nn/test_system.py]: awesome, integrated mark.parametrize into the test!

File: tests/nn/test_system.py (hunk 3 above)
Comment (ericleonardis) [reply tests/nn/test_system.py]: done!

**Review by roomrys — APPROVED** [review]

**Pull request merged**